mutatingwebhook/mutatingwebhook.go

```diff
@@ -61,48 +96,89 @@ func (mw *mutatingWebhook) handleMutate(w http.ResponseWriter, r *http.Request)
 	}
 
 	if body, err = json.Marshal(reviewResponse); err != nil {
-		log.Println(err)
+		klog.Error(err)
 		w.WriteHeader(http.StatusInternalServerError)
 		fmt.Fprintf(w, "%s", err)
 		return
 	}
 
+	w.Header().Add("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
 	w.Write(body)
 }
 
 type mutatingWebhook struct {
-	mutator Mutator
+	mutator     Mutator
+	configs     MutatingWebhookConfigs
+	server      *http.Server
+	fileWatcher *fsnotify.Watcher
 }
 
-// Starts the webserver and serves:
-// - a welcome message on /
-// - the passed Mutator on /mutate
-// - a health probe on /_healthz
-func ListenAndMutate(
+// Creates a MutatingWebhook server.
+func NewMutatingWebhook(
 	mutator Mutator,
 	configs MutatingWebhookConfigs,
-) {
+) MutatingWebhook {
+
 	configs = setDefaults(configs)
+	mux := http.NewServeMux()
+	server := http.Server{
+		Addr:           *configs.Addr,
+		Handler:        mux,
+		ReadTimeout:    *configs.ReadTimeout,
+		WriteTimeout:   *configs.WriteTimeout,
+		MaxHeaderBytes: *configs.MaxHeaderBytes,
+	}
 
-	mw := mutatingWebhook{
+	mw := &mutatingWebhook{
 		mutator: mutator,
+		configs: configs,
+		server:  &server,
 	}
 
-	mux := http.NewServeMux()
+	kpr, err := newKeypairReloader(*mw.configs.CertFilePath, *mw.configs.KeyFilePath)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	mw.fileWatcher = kpr.fileWatcher
+
+	if err := http2.ConfigureServer(mw.server, nil); err != nil {
+		klog.Fatal(err)
+	}
+
+	server.TLSConfig.GetCertificate = kpr.GetCertificateFunc()
 
 	mux.HandleFunc("/", mw.handleRoot)
 	mux.HandleFunc("/_healthz", mw.handleHealthz)
+	mux.HandleFunc("/_ready", mw.handleHealthz)
 	mux.HandleFunc("/mutate", mw.handleMutate)
 
-	s := &http.Server{
-		Addr:           *configs.Addr,
-		Handler:        mux,
-		ReadTimeout:    *configs.ReadTimeout,
-		WriteTimeout:   *configs.WriteTimeout,
-		MaxHeaderBytes: *configs.MaxHeaderBytes,
+	return mw
+}
+
+// Starts the webserver and serves:
+// - a welcome message on /
+// - the passed Mutator on /mutate
+// - a health probe on /_healthz
+// - a readiness probe on /_ready
+func (mw *mutatingWebhook) ListenAndMutate() {
```

Comment:
Do we want to just call this `ListenAndServe` rather than `ListenAndMutate`? This seems like it would be a bit clear since the function itself isn't muting, it's serving a web server?

Reply:
On it!